runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -6482,6 +6482,53 @@ TR::Register *J9::Power::TreeEvaluator::VMnewEvaluator(TR::Node *node, TR::CodeG
+
+            iCursor = generateTrg1Src1Instruction(cg, TR::InstOpCode::neg, node, offsetReg, enumReg);
+            iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::sradi, node, offsetReg, offsetReg, 63);
+            // andi_r works because sradi replicates bit 0 of offsetReg to fill the vacated positions. 1 for length > 0 and 0 otherwise.
+            iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andi_r, node, offsetReg, offsetReg, 8);
+            iCursor = generateTrg1Src1Instruction(cg, TR::InstOpCode::neg, node, offsetReg, offsetReg);
+            iCursor = generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, offsetReg, resReg, offsetReg);
+
+            iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, firstDataElementReg, offsetReg, TR::Compiler->om.discontiguousArrayHeaderSizeInBytes());
+            dataAddrSlotMR = TR::MemoryReference::createWithDisplacement(cg, offsetReg, fej9->getOffsetOfDiscontiguousDataAddrField(), TR::Compiler->om.sizeofReferenceAddress());
+            }
+         else if (!isVariableLen && node->getFirstChild()->getOpCode().isLoadConst() && node->getFirstChild()->getInt() == 0)
+            {
+            if (comp->getOption(TR_TraceCG))
+               traceMsg(comp, "Node (%p): Dealing with zero size fixed length array.\n", node);
+
+            iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, firstDataElementReg, resReg, TR::Compiler->om.discontiguousArrayHeaderSizeInBytes());
+            dataAddrSlotMR = TR::MemoryReference::createWithDisplacement(cg, resReg, fej9->getOffsetOfDiscontiguousDataAddrField(), TR::Compiler->om.sizeofReferenceAddress());
+            }
+         else
+            {
+            if (comp->getOption(TR_TraceCG))
+               traceMsg(comp, "Node (%p): Dealing with either contiguous array of non zero size or noncompressed refs variable length array.\n", node);
+
+            if (!TR::Compiler->om.compressObjectReferences())
+               TR_ASSERT_FATAL_WITH_NODE(node, fej9->getOffsetOfDiscontiguousDataAddrField() == fej9->getOffsetOfContiguousDataAddrField(), "When using full refs, dataAddr field offset is expected to be same for both discontiguous and contiguous arrays. But was %d bytes for discontiguous and %d bytes for contiguous.\n", fej9->getOffsetOfDiscontiguousDataAddrField(), fej9->getOffsetOfContiguousDataAddrField());
+
+            iCursor = generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, firstDataElementReg, resReg, TR::Compiler->om.contiguousArrayHeaderSizeInBytes());
+            dataAddrSlotMR = TR::MemoryReference::createWithDisplacement(cg, resReg, fej9->getOffsetOfContiguousDataAddrField(), TR::Compiler->om.sizeofReferenceAddress());
+            }
+
+         // store the first data element address to dataAddr slot
+         iCursor = generateMemSrc1Instruction(cg, TR::InstOpCode::stw, node, dataAddrSlotMR, firstDataElementReg);
+#endif /* TR_TARGET_64BIT */
+
```

Comment:
I'm not super-familiar with this code, but your new code maintains an instruction cursor (`iCursor`) but doesn't use it to control insertion of the instructions you generate.  Do you mean to be using the cursor-form versions of the instructions instead?  Just below here are instructions that use the cursor form.  At the very least you should be too for consistency unless there is a reason not to.

Reply:
Yeah saw this too. The `iCursor` is not being used in this new code at all, but we do have to keep it properly updated as subsequent code relies on it having the correct value. We should stay consistent with existing code and use the cursor form IMO.